### PR TITLE
Limit coordinator statemachine threads

### PR DESCRIPTION
**Pull request by pettyjamesm**
## Description

Sets a limit on the number of concurrent Query and Stage state machine callback threads that a single query have in use at once. Otherwise, queries with a large number of stages can end up triggering a very large number of native threads to be created during query failures or stage canceling events as all stages attempt to trigger callbacks at once.

Previously, `PipelinedQueryScheduler` and `EventDrivenFaultTolerantQueryScheduler` were passing `queryExecutor` into stage and output state machines which is (by default) is configured with a 1,000 thread and 1,000 queued task limits, which means that sudden bursts of state machine events could throw `RejectedExecutionException` and result in stages permanently stuck due to a state change listener callback not firing.

This PR also includes to smaller commits that:
- issues stage cancel or abort commands in topological order instead of random HashMap iteration order
- avoids pushing unnecessary submissions to the `QueryStateMachine` executor

## Release notes

(x) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQueryFactory.java
Comment (arhimondr) [inline core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQueryFactory.java:98]: Maybe simply inject the `QueryManagerConfig` here?

**Review by arhimondr — COMMENTED** [review]
The change looks good to me. Though would probably be better to get a second opinion here. Not sure who is the best fit tor take a look here, @trinodb/maintainers ?

File: core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQueryFactory.java
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQueryFactory.java]: Yeah, that is cleaner. Changed.

**Comment by pettyjamesm** [comment]
> The change looks good to me. Though would probably be better to get a second opinion here. Not sure who is the best fit tor take a look here

Maybe @findepi ? Adding as requested reviewer.

**Comment by hashhar** [comment]
FYI @raunaqmorarka @romanvainb

**Comment by pettyjamesm** [comment]
Any chance someone can take a look and give this a final review?

cc: @hashhar, @raunaqmorarka, @findepi

File: core/trino-main/src/main/java/io/trino/execution/QueryManagerConfig.java
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/execution/QueryManagerConfig.java:308]: "per-query" suffix seems redundant, maybe rename it to `query.state-machine-max-callback-threads` (e.g. `query.remote-task.max-callback-threads`).

Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/execution/QueryManagerConfig.java:308]: In this case it's not redundant: the limit is enforced on a per-query basis instead of being enforced by determining the global configuration of the the shared threadpool executor. Instead, each `QueryStateMachine` gets it's own `BoundedExecutor` to enforce the limit independent of one another.

Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/execution/QueryManagerConfig.java:308]: It appears that the rest of the configs in this file which are enforced on a per-query basis don't add the "per-query" suffix. E.g. we use query.max-stage-count and query.stage-count-warning-threshold, not query.max-stage-count-per-query and query.stage-count-warning-threshold-per-query.

Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/execution/QueryManagerConfig.java]: Fair enough, I've changed the config property name to `query.max-state-machine-callback-threads` in the latest revision.

**Review by raunaqmorarka — APPROVED** [review]

**Comment by pettyjamesm** [comment]
> @pettyjamesm is `5` enough even for more complex scenarios or it can cause concurrency throttling?

A value of 5 guarantees that each query level state machine (queryState, finalQueryInfo, and outputManager) and two stage state machines can run simultaneously without queuing. The stage state machines in particular tend to fire off in chains where one callback fires another stage to flip which submits more callbacks and previously would generate very large demand for native threads but that essentially only run for long enough to submit another callback and where queuing should actually be cheaper than a new thread creation.

That said, other than ensuring that each state machine type could run at least on at least one thread concurrently- the value 5 is essentially arbitrary. If this value is smaller than ideal, I would expect that to show up on individual queries with a large number of stages (and tasks per stage) rather than running queries concurrently, since each query is given a separate bounded executor. If anything, concurrent query execution should be more “fair” at allocating usable thread parallelism than before.

**Pull request merged**
